packages/process/process-monitor.pony

```diff
@@ -0,0 +1,366 @@
+    """
+    ifdef posix then
+      if @fcntl[I32](fd, _FSETFL(), _ONONBLOCK()) < 0 then
+        error
+      end
+    end
+  
+  be _event_notify(event: AsioEventID, flags: U32, arg: U32) =>
+    """
+    Handle the incoming event.
+    """
+    match event
+    | _stdout_event =>
+      if AsioEvent.readable(flags) then
+        _stdout_open = _pending_reads(_stdout_read)
+      elseif
+        AsioEvent.disposable(flags) then
+        @pony_asio_event_destroy(event)
+        _stdout_event = AsioEvent.none()
+      end
+    | _stderr_event =>
+      if AsioEvent.readable(flags) then
+        _stderr_open = _pending_reads(_stderr_read)
+      elseif
+        AsioEvent.disposable(flags) then
+        @pony_asio_event_destroy(event)
+        _stderr_event = AsioEvent.none()
+      end
+    end
+    _try_shutdown()
+    
+  fun ref _close() =>
+    """ 
+    Close all pipes, unsubscribe events and wait for the child process to exit.
+    """
+    ifdef posix then
+      if not _closed then
+        _closed = true
+        @close[I32](_stdin_read)
+        @close[I32](_stdin_write)
+        @close[I32](_stdout_read)
+        @close[I32](_stdout_write)
+        @close[I32](_stderr_read)
+        @close[I32](_stderr_write)
+        _stdout_open = false
+        _stderr_open = false
+        @pony_asio_event_unsubscribe(_stdout_event)
+        @pony_asio_event_unsubscribe(_stderr_event)
+      end
+      // We want to capture the exit status of the child
+      var wstatus: I32 = 0
+      let options: I32 = 0
+      if @waitpid[USize](_child_pid, addressof wstatus, options) < 0 then
+        _notifier.failed(WaitpidError)
+      end
+      // process child exit code
+      _notifier.dispose((wstatus >> 8) and 0xff)
+    end
+    
+  fun ref _try_shutdown() =>
+    """
+    If neither stdout nor stderr are open we close down and exit.
+      """
+    if _stdout_open or _stderr_open then
+      return
+    end
+    _close()
+    
+  fun ref _pending_reads(fd: U32): Bool =>
+    """
+    Read from stdout while data is available. If we read 4 kb of data, 
+    send ourself a resume message and stop reading, to avoid starving 
+    other actors.
+    It's safe to use the same buffer for stdout and stderr because of
+    causal messaging. Events get processed one _after_ another.
+    """
+    ifdef posix then
+      var sum: USize = 0
+      while true do
+        let len = @read[ISize](fd, _read_buf.cstring(), _read_buf.space())
+        let errno = @pony_os_errno()
+        let next = _read_buf.space()
+        match len
+        | -1 => match errno
+                | _EAGAIN() =>  return true // resource temporarily unavailable, retry
+                else
+                  return false
+                end
+        | 0  => match errno
+                | _EAGAIN() =>  return true // resource temporarily unavailable, retry
+                else
+                  return false
+                end
+        else
```

Comment:
I find the indenting here awkward. Minor point, of course!

How about putting the match on a new line? Then the excessive indenting goes away.